stack/maxima/contrib/linearalgebra.mac

```diff
@@ -0,0 +1,1237 @@
+ */
+symp(M):= if squarep(M) then is(M = ev(transpose(M),simp)) else false; 
+
+/** 
+ * Is a given object an invertible matrix?
+ * i.e. Is there a matrix M^^-1 such that M^^-1 . M = M . M^^-1 = I?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Is M an invertible matrix?
+ */
+invertiblep(M):= block([isInvertible],
+  isInvertible: false,
+  if squarep(M) then block(
+    if ev(is(determinant(M)#0),simp) then isInvertible: true
+  ),
+  return(isInvertible)
+);
+
+/** 
+ * Is a given object a matrix with orthogonal columns?
+ * i.e. for any two distinct columns v1 and v2, is v1 . v2 = 0?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Does M have orthogonal columns?
+ */
+orthogonal_columnsp(M):= diagp(ev(transpose(M) . M,expand,simp));
+
+/** 
+ * Is a given object a matrix with orthonormal columns?
+ * i.e. Is M^T . M = I? For columns v_i and v_j, is v_i . v_i = 1 and v_i . v_j = 0?
+ * Useful when analysing QR factorisation for rectangular matrices
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Does M have orthonormal columns?
+ */
+orthonormal_columnsp(M):= if matrixp(M) then is(ev(transpose(M).M,expand,simp) = ident(second(matrix_size(M)))) else false;
+
+/** 
+ * Is a given object an orthogonal matrix?
+ * i.e. Is M^T . M = M . M^T = I?
+ *
+ * @param[matrix] M a matrix
+ * @return[boolean] Is M an orthogonal matrix?
+ */
+orth_matrixp(M):= orthonormal_columnsp(M) and squarep(M);
+
+/*********************************************************************************/
+/* Functions to convert objects into standard forms                              */
+/*********************************************************************************/
+
+/* It is feasible that different institutions will prefer students to enter their answers in different ways */
+/* linearalgebra.mac prefers to work with either lists of lists (not distinguishing between column and row
+   vectors) or matrices whose columns are vectors of interest. */
+
+/**
+ * Takes collections of vectors and returns a list of lists.
+ * The vectors themselves may be different objects. Supported objects include:
+ * * 1xN or Mx1 matrices
+ * * Lists
+ * * Vectors using c or r notation
+ * * Sets, ntuples
+ * Other inert functions like sequence should work too, but 
+ * The collection may be a list, set, ntuple, span, or matrix
+ * If given a matrix, then returns the columns as a list, with each column also being a list.
+ *
+ * @param[] ex A list, set, ntuple or span containing vectors as matrices, lists, c, r, sets or ntuples, or a matrix
+ * @return[list] A list of lists
+ */
+make_list_of_lists(ex):= block([op1],
+  op1: safe_op(ex),
+  if not(member(op1,["[","ntuple","{","span","matrix"])) then return(ex),
+  ex: vec_convert(ex),
+  if vectorp(ex) then return([list_matrix_entries(ex)]),
+  if is(op1="matrix") then return(args(transpose(ex))),
+  ex: args(ex),
+  if ev(every(lambda([ex2],constantp(ex2) or atom(ex2)),ex),simp) then return([ex]),
```

Comment:
While working on some questions, I encountered some unexpected behavior:

I would expect `make_list_of_lists([matrix([1],[2],[3]),matrix([2],[3],[4])])` to return `[[1,2,3],[2,3,4]]`, but instead it returns 
`[[matrix([1],[2],[3]),matrix([2],[3],[4])]]`, i.e. wrapping the input into another layer of list brackets.
`make_list_of_lists([c(1,2,3),c(2,3,4)])` also yields `[[matrix([1],[2],[3]),matrix([2],[3],[4])]]`.

I observed that `constantp(matrix([1],[2],[3]))` evaluates to `true`, causing this early return. Some more investigation shows that `constantp(matrix([1],[a],[3]))` evaluates to `false`, and consequently `make_list_of_lists([matrix([1],[2],[3]),matrix([2],[a],[4])])` and `make_list_of_lists([c(1,2,3),c(2,a,4)])` return the expected output `[[1,2,3],[2,a,4]]`.

This inconsistent behavior makes me believe that `constantp/atom` is maybe not the correct check to do, and maybe a similar type check for some type of iterable like in the first two lines is more appropriate?

Some more experimentation also shows that `make_list_of_lists([1,a,g(x),4])` gives an error, because we don't do the early return here (neither constantp nor atom applies to g(x)) and the next line tries to call args on an atom.

As a more general question: What's the rationale for using lists of lists over lists of vectors (like in `subspace_equivp` and `lin_indp`) or a matrix (like `lin_indp` already supports)?

Reply:
Hi George, 

I missed this message, sorry! I did push a fix to make_list_of_lists, as I had independently run into the same issue, and also identified that `constantp` wasn't doing what I expected. The new version just checks to see whether all elements in the list are scalars, and if so return a nested list. The rationale here is that if someone called `make_list_of_lists([1,2,3])` in a linear algebra context I would personally expect the result to be `[[1,2,3]]` rather than `[1,2,3]` or `[[1],[2],[3]]`. I hope this change fixes the problems you highlighted! 

To answer the more general question: Probably a coincidence, and I'm happy for extra functionality to be added here. Originally when I was writing these, I tried to account for every reasonable input type which meant that every function had a big list of if-thens at the beginning to put everything into a form that I could work with easily. Eventually I realised I was duplicating a lot of work and decided to move that "cleaning" process into a single function. My vague memory is that the `lin_indp` function supporting matrices is because I was expecting the use case of "does this matrix have full column rank" to be common enough that it would be annoying to use `make_list_of_lists` every time. Typing this explanation out has made me feel like I undercut my previous arguments somewhat!